be/src/exec/spill/spiller.hpp

```diff
@@ -0,0 +1,156 @@
+    SCOPED_TIMER(_metrics.spill_timer);
+    RETURN_IF_ERROR(_spilled_task_status);
+    DCHECK(!chunk->is_empty());
+    DCHECK(!is_full());
+
+    if (_mem_table == nullptr) {
+        _mem_table = _acquire_mem_table_from_pool();
+        DCHECK(_mem_table != nullptr);
+    }
+
+    COUNTER_UPDATE(_metrics.spill_rows, chunk->num_rows());
+    _spilled_append_rows += chunk->num_rows();
+    TRACE_SPILL_LOG << "spilled rows:" << chunk->num_rows() << ",cumulative:" << _spilled_append_rows
+                    << ",spiller:" << this << "," << _mem_table.get();
+    RETURN_IF_ERROR(_mem_table->append(std::move(chunk)));
+    if (_mem_table->is_full()) {
+        RETURN_IF_ERROR(flush(state, std::forward<TaskExecutor>(executor), std::forward<MemGuard>(guard)));
+    }
+
+    return Status::OK();
+}
+
+template <class TaskExecutor, class MemGuard>
+Status Spiller::flush(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
+    RETURN_IF_ERROR(_spilled_task_status);
+    auto captured_mem_table = std::move(_mem_table);
+    if (captured_mem_table == nullptr) {
+        return Status::OK();
+    }
+
+    RETURN_IF_ERROR(captured_mem_table->done());
+    _running_flush_tasks++;
+    // TODO: handle spill queue
+    auto task = [this, state, guard = guard, mem_table = std::move(captured_mem_table)]() {
+        SCOPED_TIMER(_metrics.flush_timer);
+        DCHECK_GT(_running_flush_tasks, 0);
+        DCHECK(has_pending_data());
+        guard.scoped_begin();
+        //
+        auto defer = DeferOp([&]() {
+            {
+                std::lock_guard guard(_mutex);
+                _mem_table_pool.emplace(std::move(mem_table));
+            }
+
+            _update_spilled_task_status(_decrease_running_flush_tasks());
+        });
+
+        auto status_task = [state, this, mem_table]() -> Status { return _run_flush_task(state, mem_table); };
+        _update_spilled_task_status(status_task());
+        guard.scoped_end();
+    };
+    // submit io task
+    RETURN_IF_ERROR(executor.submit(std::move(task)));
+    return Status::OK();
+}
+
+template <class TaskExecutor, class MemGuard>
+StatusOr<ChunkPtr> Spiller::restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
+    SCOPED_TIMER(_metrics.restore_timer);
+    RETURN_IF_ERROR(_spilled_task_status);
+    ChunkPtr chunk;
+
+    if (_current_stream == nullptr) {
+        ASSIGN_OR_RETURN(_current_stream, _acquire_input_stream(state));
+    }
+
+    DCHECK(has_output_data());
+    // read chunk from buffer
+    ASSIGN_OR_RETURN(chunk, _current_stream->read(_spill_read_ctx));
+    TRACE_SPILL_LOG << "read rows:" << chunk->num_rows() << " cumulative:" << _restore_read_rows
+                    << ", spiller:" << this;
+    COUNTER_UPDATE(_metrics.restore_rows, chunk->num_rows());
+    _restore_read_rows += chunk->num_rows();
+
+    RETURN_IF_ERROR(trigger_restore(state, std::forward<TaskExecutor>(executor), std::forward<MemGuard>(guard)));
```

Comment:
async task generate chunks to _current_stream, and read from _current_stream if it's not empty, am I right?

Reply:
before when we call restore(). we should check spiller->has_data();